data/text_sampling/extend_tabular.py

```diff
@@ -0,0 +1,254 @@
+import glob
+import multiprocessing as mp
+import os
+import random
+import time
+from functools import partial
+
+import deepsmiles
+import pandas as pd
+import pubchempy as pcp
+import requests
+import selfies
+from rdkit import Chem
+
+# tucan needs very likely python 3.10
+# from tucan.canonicalization import canonicalize_molecule
+# from tucan.io import graph_from_molfile_text
+# from tucan.serialization import serialize_molecule
+from utils import load_yaml
+
+
+def augment_smiles(smiles: str, int_aug: int = 50, deduplicate: bool = True) -> str:
+    """
+    Takes a SMILES (not necessarily canonical) and returns `int_aug` random variations of this SMILES.
+    """
+
+    mol = Chem.MolFromSmiles(smiles)
+
+    if mol is None:
+        return None
+    else:
+        if int_aug > 0:
+            augmented = [
+                Chem.MolToSmiles(mol, canonical=False, doRandom=True)
+                for _ in range(int_aug)
+            ]
+            if deduplicate:
+                augmented = list(set(augmented))
+            return augmented
+        else:
+            raise ValueError("int_aug must be greater than zero.")
+
+
+def smiles_to_selfies(smiles: str) -> str:
+    """
+    Takes a SMILES and return the selfies encoding.
+    """
+
+    return [selfies.encoder(smiles)]
+
+
+def smiles_to_deepsmiles(smiles: str) -> str:
+    """
+    Takes a SMILES and return the DeepSMILES encoding.
+    """
+    converter = deepsmiles.Converter(rings=True, branches=True)
+    return converter.encode(smiles)
+
+
+def smiles_to_canoncial(smiles: str) -> str:
+    """
+    Takes a SMILES and return the canoncial SMILES.
+    """
+    mol = Chem.MolFromSmiles(smiles)
+    return Chem.MolToSmiles(mol)
+
+
+def smiles_to_inchi(smiles: str) -> str:
+    """
+    Takes a SMILES and return the InChI.
+    """
+    mol = Chem.MolFromSmiles(smiles)
+    return Chem.MolToInchi(mol)
+
+
+# def smiles_to_tucan(smiles: str) -> str:
+#    """
+#    Takes a SMILES and return the Tucan encoding.
+#    For this, create a molfile as StringIO, read it with graph_from_file,
+#    canonicalize it and serialize it.
+#    """
+#    molfile = Chem.MolToMolBlock(Chem.MolFromSmiles(smiles), forceV3000=True)
+#    mol = graph_from_molfile_text(molfile)
+#    mol = canonicalize_molecule(mol)
+#    return serialize_molecule(mol)
+
+
+CACTUS = "https://cactus.nci.nih.gov/chemical/structure/{0}/{1}"
+
+
+def smiles_to_iupac_name(smiles: str) -> str:
+    """Use the chemical name resolver https://cactus.nci.nih.gov/chemical/structure.
+    If this does not work, use pubchem.
+    """
+    try:
+        time.sleep(0.001)
```

Comment:
we should adjust this in a separate PR and start with a shorter sleep and then increase with exponential strategy